Data Registry views

snopoke opened the pull request:
Note this is a PR into https://github.com/dimagi/commcare-hq/pull/30011 (until that get's merged)

## Summary
This PR adds two views that allow listing, creating and updating data registries as well as managing invitations and grants

MVP Spec: https://docs.google.com/document/d/1h1chIrRkDtnPVQzFJHuB7JbZq8S4HNQf2dBA8z_MCkg/edit#heading=h.e4jelasn33v2
Mockups: https://www.figma.com/file/vzEYP2maJVa2OzrXMdqomJ/Data-Registries-Mock-ups?node-id=73%3A18

#### List view
![image](https://user-images.githubusercontent.com/249606/127622727-45d9ebad-aa1d-4150-8647-2123a5cc0afe.png)

#### Create modal
![image](https://user-images.githubusercontent.com/249606/127622801-4b4b2c0f-69b7-49ee-add8-20890f56702d.png)

#### Manage view (for owning domain)
![image](https://user-images.githubusercontent.com/249606/127844035-655e6c03-a234-4feb-8910-6dd3d916a95a.png)


#### Grants tab (owning domain)
![image](https://user-images.githubusercontent.com/249606/127835272-fbaec3f1-24a6-431f-a5c6-d91f873c37a7.png)

#### Actions tab (owning domain)
![image](https://user-images.githubusercontent.com/249606/127338695-e1d775b4-83f0-4554-9328-c411578d13ea.png)

#### Manage view (for invited domain who opted out)
![image](https://user-images.githubusercontent.com/249606/127338561-c87fb894-9bb9-43eb-9ee7-95873d9735ac.png)

#### Manage view (for invited domain who opted in)
![image](https://user-images.githubusercontent.com/249606/127844646-befd9409-ccde-4fe6-8e3d-cc6ae9004fc7.png)


## Feature Flag
`data_registry`

## Product Description
Basic views for managing Data Registries.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
Not much test coverage here. There isn't a lot of logic. I think this would mostly benefit from interaction testing which isn't something we really do. I'm open to feedback on this.

### QA Plan
QA on staging: https://dimagi-dev.atlassian.net/browse/QA-3301

### Safety story
This is all new code and new views so not much change of breaking existing features.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

snopoke committed:
automatically invite creating domain to registry

snopoke committed:
registry list static UI

snopoke committed:
make query return distinct results

snopoke committed:
remove 'badge' template

snopoke committed:
add 'new registry' button

snopoke committed:
adjust small screen sizing

snopoke committed:
add JS and convert data binding to knockout

snopoke committed:
space

snopoke committed:
div to span for labels

snopoke committed:
translate strings

snopoke committed:
nest button in H3 for alignment

snopoke committed:
use knockout modal binding

snopoke committed:
use help-block

snopoke committed:
fix JS translations

snopoke committed:
drop delete button on list + less intrusive labels

snopoke committed:
fix rejected text

snopoke committed:
move translation function to new module

snopoke committed:
convert to observables

snopoke committed:
accept invitation

snopoke committed:
add modal for accept

snopoke committed:
add reject and move actions to new JS module

snopoke committed:
cleanup + remove invite button from list

snopoke committed:
plumbing for edit / manage

snopoke committed:
edit view v0.1

snopoke committed:
remove project space

snopoke committed:
add projects

snopoke committed:
add params to docs

snopoke committed:
move invitation management to separate endpoint

snopoke committed:
put new invitations at the top

snopoke committed:
add sort icon

snopoke committed:
update remove button

snopoke committed:
fixup after rebase

snopoke committed:
fix breadcrumbs etc

snopoke committed:
make schema editing 'inline' with dialog

snopoke committed:
update grants list

snopoke committed:
create grants

snopoke committed:
remove grants

snopoke committed:
activate / deactivat + styling

snopoke committed:
fix model binding

snopoke committed:
active state in list view

snopoke committed:
click handlers

snopoke committed:
manage unowned registries

snopoke committed:
fix modal in list

snopoke committed:
opt in / out + don't show grants unless accepted

snopoke committed:
make notice an info box

snopoke committed:
add note about owner

snopoke committed:
get case types from DD

snopoke committed:
add more help text

snopoke committed:
delete registry

snopoke committed:
saving indicators

snopoke committed:
tidy up

snopoke committed:
get domains from billing account

snopoke committed:
add list of participating project spaces

snopoke committed:
add logging

snopoke committed:
create CRUD helper and update tests

snopoke committed:
add toggle required decorator + change view order

snopoke committed:
remove unused method

snopoke committed:
lint etc

orangejenny commented:
not going to get to this today, should be able to do it tomorrow or Friday

snopoke committed:
Merge branch 'master' into sk/data-registry-view

snopoke committed:
redo how modal if works

snopoke committed:
use 'static' from hq_shared_tags

snopoke committed:
move registry to project settings

snopoke committed:
switch domain selects to use 'multiselect'

snopoke committed:
support function based view

snopoke committed:
filter out granted domains

snopoke committed:
only show selector if there are items to select

snopoke committed:
create registry

snopoke committed:
update button text

orangejenny commented on corehq/apps/registry/templates/registry/registry_edit.html:
```diff
@@ -408,4 +435,82 @@ <h4 class="modal-title">{% trans "Remove Grant" %}</h4>
+                                     'fa-spin': formDeleteRegistrySent
+                                 }
+                             "></i>
+                {% trans "Delete Mobile Worker" %}
```
"Delete Registry"

orangejenny commented on corehq/apps/registry/templates/registry/registry_edit.html:
```diff
@@ -448,7 +452,14 @@ <h4 class="modal-title">{% trans "Grant access" %}</h4>
+                selectableHeaderTitle: '{% trans_html_attr "Available Project Spaces" %}',
+                selectedHeaderTitle: '{% trans_html_attr "Selected" %}',
+                searchItemTitle: '{% trans_html_attr "Search" %}',
+                values: availableGrantDomains
```
@snopoke This makes the multiselect accept dynamic values, right?

@gherceg FYI, since you've also been working on this widget.

snopoke replied:
yes that's correct, it can now accept values from a knockout observable. I haven't tested adding / removing values dynamically though.

snopoke commented on corehq/apps/hqwebapp/static/registry/less/registry.less:
```diff
@@ -17,3 +22,29 @@
+}
+
+// HELPERS
+// These helpers are configured to match the defaults in Bootstrap 4
```
they utilities are the same between 4 & 5 but I'll update the comment

snopoke committed:
lint

snopoke committed:
update BS comment

snopoke committed:
use trans_html_attr

snopoke committed:
translate strings

snopoke committed:
add help text to case types select box

snopoke committed:
send invitation email

snopoke committed:
add registry signals

snopoke committed:
send invitation email

snopoke committed:
test more logging

snopoke committed:
list grants to me above tabs

snopoke committed:
don't show all grants by default

orangejenny commented on corehq/apps/domain/tests/test_utils.py:
```diff
@@ -32,3 +34,17 @@ def test_guess_domain_language(self):
+        yield domain
+    finally:
+        if skip_full_delete:
+            Domain.get_db().delete_doc(domain.get_id)
```
How do you know whether or not it's safe to only delete the doc? I'm thinking less about this PR and more about all the other tests where we do the full `domain.delete()`.

snopoke replied:
If you have created other couch docs that you aren't cleaning up manually then you'll need to do a full delete. In most cases there are only a few docs that are being created by tests and it would be much faster to delete them manually than to rely on `domain.delete`

snopoke committed:
Merge branch 'master' into sk/data-registry-view

snopoke committed:
display correct domain as owner

snopoke committed:
Merge branch 'sk/data-registry-signals' into sk/data-registry-view

snopoke committed:
fix alert message

snopoke committed:
update observable on input change

snopoke committed:
correctly call signal

snopoke committed:
consistent error handling

snopoke committed:
fix var + lint

snopoke committed:
fix kwarg name

snopoke committed:
fix kwargs typos

snopoke committed:
pluralize invitations sent message

snopoke committed:
set response date to blank for pending invitations

snopoke committed:
make email link absolute

snopoke committed:
make registry slug globally unique

slug must be unique since domains could have access to multiple registries
with the same slug which causes issues with the views. This could
be overcome but it's much simpler to make the slug globally unique.

snopoke committed:
fix blank owner in template

snopoke committed:
fix keyerror

snopoke committed:
update grant modal text

snopoke committed:
fix and update display of participating project spaces

snopoke committed:
always show list of participating domains & exclude current domain
if invitation is not accepted

snopoke committed:
add help text with link to docs

snopoke committed:
Merge branch 'master' into sk/data-registry-view

snopoke committed:
allow specifying domain name

snopoke committed:
fix uniqueness test

snopoke committed:
make opt in button the same color as opt out

snopoke committed:
send invitation response email

snopoke committed:
add grant email

snopoke committed:
fix name in email text

snopoke committed:
subdevide email text

snopoke committed:
fix signal kwargs

snopoke committed:
don't send invitation response email for owning domain

esoergel commented on corehq/apps/registry/models.py:
```diff
@@ -26,6 +27,17 @@ def owned_by_domain(self, domain, is_active=None):
             query = query.filter(is_active=is_active)
         return query
 
+    def visible_to_domain(self, domain):
```
When would you want to do `visible_to_domain(domain)` vs `accessible_to_domain(domain)`?  Why have both?  Looks like the difference is that the latter will exclude a domain's own registries unless its invited.  Are projects ever, sometimes, or never invited to their own registries?

snopoke replied:
The distinction comes in when you want to access the data in a registry. In order opt in / out of the registry you need to be able to 'see' it in the UI so that's 'visible_to_domain'. This is also likely what will get used in app manager / UCRs when configuring case search etc.

If you want to query the registry then you need to have accepted the invitation and the registry must be 'active' and that's covered by 'accessible_to_domain'. 

If you have better naming ideas I'm all ears. I'll add a bit more to the doc string as well.

esoergel replied:
no that makes sense, thanks

snopoke committed:
remove duplicate create method

snopoke committed:
update migration

snopoke committed:
put invitation in context

snopoke committed:
use correct domain for url

snopoke committed:
move comment

snopoke committed:
make case types required when creating a registry

snopoke committed:
check name uniqueness when creating registry

snopoke committed:
validate params during create

snopoke committed:
validate name during edit

snopoke committed:
Merge branch 'master' into sk/data-registry-view

snopoke committed:
lint

snopoke commented on corehq/apps/registry/static/registry/js/registry_list.js:
```diff
@@ -0,0 +1,101 @@
+
+        // CREATE workflow
+        self.name = ko.observable("").extend({
+            rateLimit: { method: "notifyWhenChangesStop", timeout: 400, }
```
Gee thanks @stickler-ci - that's not confusing at all....

snopoke committed:
disable save button if name validation fails

snopoke committed:
switch to master branch version of multiselect

snopoke committed:
reset selected grants when closing modal

snopoke committed:
use correct domain name in invitation deletion message

snopoke committed:
fix registry migration

snopoke committed:
Merge branch 'master' into sk/data-registry-view

snopoke committed:
fix slugs unique

snopoke commented:
@esoergel @orangejenny QA has passed. I'm going to merge this in. If there is additional feedback I can address it in followup PRs.
